src/core/cgroup.c

```diff
@@ -306,30 +307,36 @@ void cgroup_context_dump(CGroupContext *c, FILE* f, const char *prefix) {
         }
 }
 
-static int lookup_block_device(const char *p, dev_t *dev) {
+static int lookup_block_device(const char *p, dev_t *ret) {
         struct stat st;
+        int r;
 
         assert(p);
-        assert(dev);
+        assert(ret);
 
         if (stat(p, &st) < 0)
-                return log_warning_errno(errno, "Couldn't stat device %s: %m", p);
+                return log_warning_errno(errno, "Couldn't stat device '%s': %m", p);
 
         if (S_ISBLK(st.st_mode))
-                *dev = st.st_rdev;
-        else if (major(st.st_dev) != 0) {
-                /* If this is not a device node then find the block
-                 * device this file is stored on */
-                *dev = st.st_dev;
-
-                /* If this is a partition, try to get the originating
-                 * block device */
-                (void) block_get_whole_disk(*dev, dev);
-        } else {
-                log_warning("%s is not a block device and file system block device cannot be determined or is not local.", p);
-                return -ENODEV;
+                *ret = st.st_rdev;
+        else if (major(st.st_dev) != 0)
+                *ret = st.st_dev; /* If this is not a device node then use the block device this file is stored on */
+        else {
+                /* If this is btrfs, getting the backing block device is a bit harder */
+                r = btrfs_get_block_device(p, ret);
+                if (r == -ENOTTY) {
+                        log_warning("'%s' is not a block device node, and file system block device cannot be determined or is not local.", p);
+                        return -ENODEV;
+                }
+                if (r < 0)
+                        return log_warning_errno(r, "Failed to determine block device backing btrfs file system '%s': %m", p);
```

Comment:
This message assumes that `p` must be a btrfs device. I think it'd be better to reorder this and report the more specific case of failing to determine btrfs backing device first:
```C
                r = btrfs_get_block_device(p, ret);
                if (r < 0 && r != -ENOTTY)
                        return log_warning_errno(r, "Failed to determine if path '%s' is backed by a btrfs file system: %m", p);

                if (r == -ENOTTY) {
                        log_warning("'%s' is not a block device node, and file system block device cannot be determined or is not local.", p);
                        return -ENODEV;
                }
```